Update commands used for TS in troubleshoot-network-aks.md (#4402)
MicrosoftDocs/architecture-center: docs/operator-guides/aks/troubleshoot-network-aks.md

delegate: Failed to allocate address: No available addresses 
```

Or a `not enough IPs available` error:

```output
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox 
'ac1b1354613465324654c1588ac64f1a756aa32f14732246ac4132133ba21364': plugin type='azure-vnet' 
failed (add): IPAM Invoker Add failed with error: Failed to get IP address from CNS with error: 
%w: AllocateIPConfig failed: not enough IPs available for 9c6a7f37-dd43-4f7c-a01f-1ff41653609c, 
waiting on Azure CNS to allocate more with NC Status: , IP config request is [IPConfigRequest: 
DesiredIPAddress , PodInterfaceID a1876957-eth0, InfraContainerID 
a1231464635654a123646565456cc146841c1313546a515432161a45a5316541, OrchestratorContext 
{'PodName':'a_podname','PodNamespace':'my_namespace'}]
```

> 

Reviewer: ```suggestion
```

delegate: Failed to allocate address: No available addresses 
```

> Another example of the error:

Reviewer: ```suggestion
Or a `not enough IPs available` error:
```

```output
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox 
'ac1b1354613465324654c1588ac64f1a756aa32f14732246ac4132133ba21364': plugin type='azure-vnet' 
failed (add): IPAM Invoker Add failed with error: Failed to get IP address from CNS with error: 
%w: AllocateIPConfig failed: not enough IPs available for 9c6a7f37-dd43-4f7c-a01f-1ff41653609c, 
waiting on Azure CNS to allocate more with NC Status: , IP config request is [IPConfigRequest: 
DesiredIPAddress , PodInterfaceID a1876957-eth0, InfraContainerID 
a1231464635654a123646565456cc146841c1313546a515432161a45a5316541, OrchestratorContext 
{'PodName':'a_podname','PodNamespace':'my_namespace'}]
```

Check the allocated IP addresses in the plugin IPAM store. You might find that all IP addresses are allocated, but the number is much less than the number of running Pods:

> **If using [`kubenet`](/azure/aks/configure-kubenet):**

Reviewer: ```suggestion
**If using [kubenet](/azure/aks/configure-kubenet):**
```

```bash
# Kubenet, for example. The actual path of the IPAM store file depends on network plugin implementation. 
chroot /host/
ls -la "/var/lib/cni/networks/$(ls /var/lib/cni/networks/ | grep -e "k8s-pod-network" -e "kubenet")" | grep -v -e "lock\|last\|total" -e '\.$' | wc -l
244
```

> [!NOTE]

> > For `kubenet` without `calico`, the path is `/var/lib/cni/networks/kubenet`. For `kubenet` with `calico`, the path is `/var/lib/cni/networks/k8s-pod-network`. In this demo, it will auto-select the path while executing the command.

Reviewer: ```suggestion
> For kubenet without Calico, the path is `/var/lib/cni/networks/kubenet`. For kubenet with Calico, the path is `/var/lib/cni/networks/k8s-pod-network`. The script above will auto select the path while executing the command.
```

```bash
# Check running Pod IPs
kubectl get pods --field-selector spec.nodeName=<your_node_name>,status.phase=Running -A -o json | jq -r '.items[] | select(.spec.hostNetwork != 'true').status.podIP' | wc -l
7 
```

> **If using [`Azure CNI for dynamic IP allocation`](/azure/aks/configure-azure-cni-dynamic-ip-allocation):**

Reviewer: ```suggestion
**If using [Azure CNI for dynamic IP allocation](/azure/aks/configure-azure-cni-dynamic-ip-allocation):**
```

> ```

Reviewer: ~~~suggestion
```bash
~~~

kubectl get nnc -n kube-system -o wide

> NAME                               REQUESTED IPS  ALLOCATED IPS  SUBNET  SUBNET CIDR   NC ID                                 NC MODE  NC TYPE  NC VERSION

Reviewer: ~~~suggestion
```

```output
NAME                               REQUESTED IPS  ALLOCATED IPS  SUBNET  SUBNET CIDR   NC ID                                 NC MODE  NC TYPE  NC VERSION
~~~

aks-agentpool-12345678-vmss000000  32             32             subnet  10.18.0.0/15  559e239d-f744-4f84-bbe0-c7c6fd12ec17  dynamic  vnet     1
```

> ```

Reviewer: ~~~suggestion
```bash
~~~

```bash
# Kubenet, for example. The actual path of the IPAM store file depends on network plugin implementation. 
chroot /host/

> cd /var/lib/cni/networks/k8s-pod-network

Reviewer: The comment says for "kubenet" -- just to verify, this location is for kubnet?  If not, please update the comment.

Author: @ckittel I created two `kubenet` AKS and they are eventually creating this folder. And the folder `kubnet` (which is used by current document) is not existing. 
I did not verify if this is the folder also for `normal CNI`, `CNI overlay`. But I checked it is not for `CNI with Podsubnet`. Or maybe the behaviour will be different if not using custom VNet (although I don't think there will be any). I will further verify later and update the comment.

Reviewer: Sounds good.  If you had it on a kubenet cluster, that's good enough for me.  I'll mark it resolved.  If you find something that should require a change with this though, just push it with your next push :)

ls -la | grep -v -e "lock\|last" -e '\.$' | wc -l
258 

> ```
> ```bash

Reviewer: ```suggestion

```

# Check running Pod IPs

> kubectl get pods --field-selector spec.nodeName=<your_node_name>,status.phase=Running -A -o json | jq -r '.items[] | select(.spec.hostNetwork != 'true').status.podIP' | wc -l

Reviewer: This section is specifically for ALL IP addresses allocated, not just the running ones

See the text above that says:

> Check the allocated IP addresses in the plugin IPAM store. You might find that all IP addresses are allocated, but the number is much less than the number of running Pods:

I feel like you're changing the intent of this section.  I appreciate the updates, but can you make sure the new commands list ALL allocated IP addresses, not just those that are running.

If you want to add a "running only" varient, that's fine.  But let's keep the original task being solved for (with latest commands).

Author: @ckittel My Apology. When I trying to find out the correct way to use `jq`, I get lost and forget the purpose of this part. But to fix this part, I have to also exclude "Completed" Pods which will also returned with an IP. 
I will verify the new command (that I will redesign) and leave you the comment once I checked the document (and do the lab) and see there is no other condition will get "Pod IPs will released IPs".  
My work time will start in 6 hours, so I will get back in 8–10 hours.  

Thank you for your reviewing and pointing out my mistake.

Author: @ckittel Wait, I just realised I may not be wrong here.
The all allocated IP by `kubenet` is from the command `ls -la /var/lib/cni/networks/k8s-pod-network`, this line is supposed to get the running Pod IPs.

Original statement: `Check the allocated IP addresses in the plugin IPAM store. You might find that all IP addresses are allocated, but the number is much less than the number of running Pods:`
Get allocated IPs: `ls -la /var/lib/cni/networks/k8s-pod-network (result: 258)`
Get actual running PodsIPs: `kubectl get pods --field-selector spec.nodeName=<your_node_name>,status.phase=Running -A -o json | jq -r '.items[] | select(.spec.hostNetwork != 'true').status.podIP' | wc -l (result: 7)`

So I am not challenging the section at all?

Reviewer: Let me ask this, what is the mechanism to get all of the "allocated IP addresses in the network plugin's IPAM store?"

If you can demonstrate that you can
- Generate a list of IP addresses are currently marked as "allocated" such that exhaustion of those IPs would result in a `Failed to allocate address` error.  Then this can merge in.

If

```bash
chroot /host/
cd /var/lib/cni/networks/k8s-pod-network
ls -la | grep -v -e "lock\|last" -e '\.$' | wc -l
```

is that. Then, correct.  I was tilting at windmills here.  Please confirm that, and we can proceed.

(BTW, thank you for your diligence on this!)

Author: @ckittel After consulting technical advisor, it is found that no matter how you see it, there is no Kubernetes API will return what IPs are allocated. You cannot even see any code can indicate how many IPs are being used:
https://github.com/kubernetes/kubernetes/blob/2ec63e0d28951bb525a5bce0d9459afa1c71c0bd/pkg/controller/nodeipam/node_ipam_controller.go

There is a very high chance that  `ls -la /var/lib/cni/networks/(k8s-pod-network|kubenet)` is the only way to see how many IPs are being allocated by IPAM. Because it generates a filed name with IP.  
(BTW: this is also the method what used in the current version.)

There is also another way to modify this document: we use `Podsubnet` AKS instead. It will be much easier to check. For example:  
```
kubectl get nnc -n kube-system -o wide
NAME                               REQUESTED IPS  ALLOCATED IPS  SUBNET            SUBNET CIDR   NC ID                                 NC MODE  NC TYPE  NC VERSION
aks-agentpool-12345678-vmss000000  32             32             aks-pod-subnet-1  10.18.0.0/15  559e239d-f744-4f84-bbe0-c7c6fd12ec17  dynamic  vnet     1
```
 Then checking current running Pods:
```
kubectl get pods --field-selector spec.nodeName=aks-agentpool-12345678-vmss000000,status.phase=Running -A -o json | jq -r '.items[] | select(.spec.hostNetwork != 'true').status.podIP' | wc -l
```

In the past, I faced a user whose allocated IPs and requested IPs are both 32, but running Pods only has 21 (or what else) but cannot create any more Pods. 
```
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox 'ac1b1354613465324654c1588ac64f1a756aa32f14732246ac4132133ba21364': plugin type='azure-vnet' failed (add): IPAM Invoker Add failed with error: Failed to get IP address from CNS with error: %w: AllocateIPConfig failed: not enough IPs available for 9c6a7f37-dd43-4f7c-a01f-1ff41653609c, waiting on Azure CNS to allocate more with NC Status: , IP config request is [IPConfigRequest: DesiredIPAddress , PodInterfaceID a1876957-eth0, InfraContainerID a1231464635654a123646565456cc146841c1313546a515432161a45a5316541, OrchestratorContext {'PodName':'a_podname','PodNamespace':'my_namespace'}]
```

It is up to you to choose which one you want to use. I will make final push based on your preference.

Reviewer: Let's include both, the "raw" `ls -la /path` version(s), and the Kube API version.

Then we can include a "running pod" IP dump (like you have) just so a user can see the disconnect between the number of running pods vs the allocated IPs.

Have a great weekend!

Author: I have submitted my next push. Kindly check it when you are available.
